### DISCUSSION TITLE
guides/docker-quickstart/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8719] [Docs]
# guides/docker-quickstart/

Complete guide to setting up and using Ultralytics YOLO models with Docker. Learn how to install Docker, manage GPU support, and run YOLO models in isolated containers.

https://docs.ultralytics.com/guides/docker-quickstart/

**Comment by udayzee05** [comment]
DEPRECATION NOTICE
The nvidia-docker wrapper is no longer supported, and the NVIDIA Container Toolkit has been extended to allow users to configure Docker to use the NVIDIA Container Runtime. For further instructions, see the NVIDIA Container Toolkit documentation and specifically the install guide.22 Jan 2024

**Reply by pderrenger** [reply]
Hey there! 👋 Thanks for bringing up the deprecation notice regarding the nvidia-docker wrapper. You're absolutely right; it's important to switch to the NVIDIA Container Toolkit for GPU support in Docker containers now. Here's a quick snippet on how to set it up, based on our Docker Quickstart Guide:

```bash
# Add NVIDIA package repositories
curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
distribution=$(lsb_release -cs)
curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list

# Install NVIDIA Docker runtime
sudo apt-get update
sudo apt-get install -y nvidia-docker2

# Restart Docker service to apply changes
sudo systemctl restart docker
```

And to verify the NVIDIA runtime with Docker:
```bash
docker info | grep -i runtime
```

This should show `nvidia` in the list of runtimes, indicating that your Docker setup can now leverage GPU acceleration. 🚀

For more detailed instructions and tips on using Ultralytics with Docker, including mounting local directories and running containers, check out our [Docker Quickstart Guide](https://docs.ultralytics.com/guides/docker-quickstart/). Happy coding!

**Comment by Peter-coding** [comment]
Hey!
I trained a YOLOv8 model on my local computer. I have the best.pt file, and I can use it in CLI (conda) for predict. I want to make a python app, which uses this best.pt file, and I want to containerize it. Is it possible to write the full python app (it is just a few lines of code, which runs the detection, then writes the detection result to a database), then containerize it, and able to use it anywhere?
Thank you! :D

**Reply by pderrenger** [reply]
Hey there! 😊

Absolutely, you can absolutely containerize your Python app that uses the `best.pt` model! Here's a quick guide:

1. **Create Your Python App**: Just like you mentioned, write your Python code that performs the detection and interacts with the database. 

2. **Dockerize Your Application**: 
   - **Dockerfile**: Create a Dockerfile in your project directory. Here's a simple example that uses Ultralytics' Docker image as a base:

     ```Dockerfile
     FROM ultralytics/ultralytics:latest
     WORKDIR /app
     COPY . /app
     RUN pip install -r requirements.txt  # Assuming you have a requirements.txt
     CMD ["python", "your_script.py"]
     ```
   - **Build Your Docker Image**: Run the following command in your project directory:
     ```bash
     sudo docker build -t your-app-name .
     ```
   - **Run Your Container**: To run your container, especially if you're planning to use GPU, you might want to specify `--gpus all`:
     ```bash
     sudo docker run -it --ipc=host --gpus all your-app-name
     ```
     For accessing local files, consider mounting a volume as described in the Docker Quickstart Guide.

3. **Deployment**: With your Docker image built, you can now deploy and run it anywhere that Docker is supported. If using a GPU, ensure the host machine supports NVIDIA GPUs and has the necessary drivers and Docker configured correctly.

That's it! You should be able to use your model within a containerized app now. If you hit any snags, double-check the [Docker Quickstart Guide for Ultralytics](https://docs.ultralytics.com/guides/docker-quickstart/) for more details. Happy coding! 😄

**Comment by HWKevinhwk** [comment]
I want to install ultralytics using docker on top of NVIDIA TX2, is there a detailed procedure? The reason for this question is that after entering the command nvidia-smi in the TX2 terminal, the output is command not found, and I'm worried that the steps in the document above are not applicable to TX2.

**Reply by pderrenger** [reply]
@HWKevinhwk hey there! 🚀 Setting up Ultralytics on NVIDIA TX2 with Docker can be a bit tricky if `nvidia-smi` is missing, but no worries, I can guide you through a process that should work.

For NVIDIA Jetson platforms like TX2, `nvidia-smi` might not be available, and you'll typically use `tegrastats` instead to monitor GPUs. However, for Docker integration, you're on the right path with our Docker Quickstart Guide. You'll want to use our Jetson-optimized Docker image.

Before anything, ensure you've installed [JetPack](https://developer.nvidia.com/embedded/jetpack) as it includes the NVIDIA drivers and libraries needed for GPU acceleration.

Then, instead of `nvidia-docker2`, Jetson devices work with the Docker integration provided through JetPack. You can then proceed to pull and run the Ultralytics Docker image specifically designed for Jetson devices:

```bash
# Pull Ultralytics image for Jetson
docker pull ultralytics/ultralytics:jetson-latest

# Run the Docker container
docker run -it --runtime nvidia --network host ultralytics/ultralytics:jetson-latest
```

Note: The key difference is using `--runtime nvidia` (provided by JetPack's Docker integration) instead of `--gpus all` which applies to desktop GPUs with `nvidia-docker2`.

This should set you up for running Ultralytics smoothly on your TX2! If you encounter any specific issues, feel free to drop them here. Happy detecting! 🎉

**Reply by HWKevinhwk** [reply]
Is there a detailed tutorial on deploying yolov8 on jetson TX2 using docker? I am pulling images docker pull ultralytics/ultralytics:latest-jetson terminal will downloading pause, long time pulling images unsuccessful, repeated attempts still so, any solution please?

**Reply by pderrenger** [reply]
@HWKevinhwk hey there! 😊

It sounds like you're facing issues with the Docker image pull for the Jetson TX2. This might be due to network issues or Docker Hub delays.

Here’s a simplified breakdown to get you started with deploying YOLOv8 on the TX2 using Docker:

1. **Ensure JetPack is installed**: This includes CUDA-compatible GPU drivers necessary for Docker integration with the GPU on the Jetson TX2.

2. **Install Docker**:
   ```bash
   sudo apt-get install docker.io
   ```

3. **Setup NVIDIA Docker support** (JetPack should handle this):
   Just double-check that you have the nvidia-docker plugin. If not, you might need to install it as per JetPack’s documentation because `nvidia-smi` doesn't work on Jetson.

4. **Pull the Optimized Docker Image for Jetson**:
Remember, pulling large Docker images can sometimes be challenging on slower networks.
   ```bash
   docker pull ultralytics/ultralytics:latest-jetson
   ```

   If you keep facing issues while pulling the image, try restarting your Docker service or check your network settings. Sometimes, using a different network can help too!

5. **Run the Image**:
   ```bash
   docker run -it --runtime nvidia --network host ultralytics/ultralytics:latest-jetson
   ```

This should set you on the right path to running YOLOv8 on your Jetson TX2. If the problem persists, it might be worth checking Docker’s log or discussing this issue further with your network admin or on Docker forums! 😄

Happy coding! 🚀

**Comment by HellaPinot** [comment]
Hello! 
I have been testing running yolov8 inside a container, things seem to work well however after a somewhat random amount of time the inference hangs with no error output.

I know the container hasn't crashed as I am reading events from a processing queue and I can see new partitions are being allocated on another thread.

I'm running two models one after the other (different sizes and detection classes), so I'm wondering if this could be the root of the issue? At no point would they be running at the same time.

I believe it is happening exactly when I call predict. Its hard to debug as again it happens very randomly and there is not output from the container, it seems like there might be an infinite loop somewhere thats stopping it from continuing.

Curious if anything like this has been encountered before?

**Reply by glenn-jocher** [reply]
Hello! 👋

It sounds like a tricky issue, especially since it occurs randomly. Here are a few suggestions that might help you isolate or resolve the problem:

1. **Logging and Monitoring:** Enhance the logging within your container. Consider logging the input sizes and other relevant data before the prediction call, and make sure you're logging any exceptions that could be silently caught.

2. **Resource Checks:** Ensure that there are no resource starvation issues, such as GPU memory not being freed up properly. Tools like `nvidia-smi` can help monitor GPU usage.

3. **Sequential Model Loading:** Although the models are run sequentially and not in parallel, it could be beneficial to explicitly clear the memory after each model's predictions are done. This can be done by deleting model objects and calling `torch.cuda.empty_cache()` if you are using PyTorch.

4. **Hardware Testing:** Run your container pointing the inference to CPU only for a time to see if the problem persists. This could help determine if the issue is related to GPU computation or not.

5. **Reproducibility Tactic:** Since the issue is sporadic, try to determine if there’s a specific type of input or specific conditions under which this occurs more frequently.

If these suggestions don't help, please provide more details or code snippets, and I’d be happy to dive deeper into this!

Let me know how it goes!

**Comment by lauravchrobak** [comment]
It looks like the, "Running Ultralytics in Docker Container" section is duplicated here

**Reply by glenn-jocher** [reply]
@lauravchrobak thank you for pointing that out. We'll review and correct the duplication in the documentation. If you have any other questions or need further assistance, feel free to ask!

**Comment by leticiazorzirama** [comment]
Hello,

Is this workflow suitable for other YOLO families, such as YOLO26?

**Reply by glenn-jocher** [reply]
Yes, the same workflow supports YOLO26 and other model families available in the `ultralytics` package—use an up-to-date official image and specify the appropriate weights, such as `model=yolo26n.pt`. See the [Docker Quickstart Guide](https://docs.ultralytics.com/guides/docker-quickstart) for current image options, and feel free to drop any follow-up questions here so the Ultralytics team can help.

**Reply by leticiazorzirama** [reply]
Right Glenn, thanks
